back-end/server.py

```diff
@@ -0,0 +1,120 @@
+import logging
+import os
+from flask import Flask, request, redirect, send_from_directory
+from werkzeug.utils import secure_filename
+
+import convert
+
+log = logging.getLogger("main")
+logging.basicConfig(level=os.environ.get("LOGLEVEL", "INFO"))
+
+app = Flask(__name__)
+app.config['MAX_CONTENT_LENGTH'] = 16 * 1024 * 1024
+app.config['ALLOWED_EXTENSIONS'] = set(
+    ['pdf', 'png', 'jpg', 'jpeg', 'jpg_large'])
+app.config['SUPPORTED_CONVERSIONS'] = {
+    'jpg': ['jpg', 'pdf', 'png'],
+    'jpg_large': ['jpg', 'pdf', 'png'],
+    'jpeg': ['jpg', 'pdf', 'png'],
+    'pdf': ['pdf'],
+    'png': ['jpg', 'pdf', 'png']
+}
+app.config['CONVERT'] = {
+    'jpg': convert.to_jpg,
+    'pdf': convert.to_pdf,
+    'png': convert.to_png
+}
+
+app.config['UPLOAD_FOLDER'] = 'uploads'
+app.config['UPLOAD_FOLDER_PATH'] = os.path.join(
+    app.root_path, app.config['UPLOAD_FOLDER'])
+
+
+def get_extension(filename):
+    _, e = os.path.splitext(filename)
+    return e.lower()[1:]
+
+
+def check_extension(filename):
+    ext = get_extension(filename)
+    return ext in app.config['ALLOWED_EXTENSIONS']
+
+
+def allowed_file(filename):
+    return '.' in filename and check_extension(filename)
+
+
+def check_supported_conversion(output, file):
+    input = get_extension(file)
+    if output not in app.config['SUPPORTED_CONVERSIONS'][input]:
+        raise ValueError("Conversion not supported")
+
+
+def upload(file):
+    if not allowed_file(file.filename):
+        raise ValueError("File not supported")
+    filename = secure_filename(file.filename)
+    filepath = os.path.join(app.config['UPLOAD_FOLDER_PATH'], filename)
+    log.info("Saving to {}".format(filepath))
+    file.save(filepath)
+    return filepath
+
+
+@app.route('/', methods=['GET', 'POST'])
+def upload_file():
+    if request.method == 'POST':
+        if 'file' not in request.files:
+            return 'No file part', 204
+        file = request.files['file']
+
+        if file.filename == '':
+            return 'No selected file', 204
+
+        try:
+            log.info("About to upload {}".format(file))
+            input_filepath = upload(file)
+            requested_conversion = request.form.get('output')
+            check_supported_conversion(requested_conversion, input_filepath)
+            converter = app.config['CONVERT'][requested_conversion]
```

Comment:
I wonder if it might be easier to read/understand this flow if the converter only performed the convert action and returned the resulting `Image`, leaving this HTTP handler to perform the actual `.save`? Otherwise one has to dive deeper to understand that these converters have side effects.

As a bonus, this would generalize much of the error handling currently spread around the converters.

This is a nontrivial change so feel free to ignore it, but it might make it easier to explain the code to fresh eyes `¯\_(ツ)_/¯`

Reply:
Unfortunately I'd then have to put logic here to differentiate between the files that get passed back (pdf's are saved differently than images)

Reply:
Or is there something I am missing?

Reply:
The PDF flow is a bit different from the rest, so yea, it would still need special treatment. Possibly not worth it for something that's just an example app.